mne/tests/test_morph.py

```diff
@@ -0,0 +1,258 @@
+
+    stc_vol = read_source_estimate(fname_vol, 'sample')
+
+    # check for invalid input type
+    pytest.raises(ValueError, SourceMorph, src=42)
+
+    # check for raising an error if neither
+    # inverse_operator_vol['src'][0]['subject_his_id'] nor subject_from is set,
+    # but attempting to perform a volume morph
+    src = inverse_operator_vol['src']
+    src[0]['subject_his_id'] = None
+    pytest.raises(ValueError, SourceMorph, src=src, subjects_dir=subjects_dir)
+
+    # check infer subject_from from src[0]['subject_his_id']
+    src[0]['subject_his_id'] = 'sample'
+    source_morph_vol = SourceMorph(subjects_dir=subjects_dir,
+                                   src=inverse_operator_vol['src'],
+                                   niter_affine=(10, 10),
+                                   niter_sdr=(3, 3), spacing=7)
+
+    assert source_morph_vol.subject_from == 'sample'
+
+    # the brain used in sample data has shape (255, 255, 255)
+    assert (tuple(
+        source_morph_vol.params['DiffeomorphicMap']['domain_shape']) ==
+        (37, 37, 37))
+
+    assert (tuple(source_morph_vol.params['AffineMap']['domain_shape']) ==
+            (37, 37, 37))
+
+    # proofs the above
+    assert source_morph_vol.spacing == 7
+
+    # assure proper src shape
+    assert source_morph_vol.params['src_shape_full'] == (
+        src[0]['mri_height'], src[0]['mri_depth'], src[0]['mri_width'])
+
+    # check full mri resolution registration (takes very long)
+    # source_morph_vol = SourceMorph(
+    #     os.path.join(sample_dir, 'sample_audvis-meg-vol-7-fwd.fif'),
+    #     subject_from='sample',
+    #     subjects_dir=os.path.join(data_path, 'subjects'),
+    #     grid_spacing=None)
+
+    fwd = read_forward_solution(
+        os.path.join(sample_dir, 'sample_audvis-meg-vol-7-fwd.fif'))
+    # check input via path to src and path to subject_to
+    source_morph_vol = SourceMorph(
+        subject_from='sample',
+        subject_to=os.path.join(data_path, 'subjects', 'fsaverage', 'mri',
+                                'brain.mgz'),
+        subjects_dir=subjects_dir, niter_affine=(10, 10),
+        src=fwd['src'],
+        niter_sdr=(3, 3), spacing=7)
+
+    # check wrong subject_to
+    pytest.raises(IOError, SourceMorph,
+                  subject_from='sample', subject_to='42',
+                  subjects_dir=subjects_dir,
+                  src=fwd['src'])
+
+    # two different ways of saving
+    source_morph_vol.save(os.path.join(tempdir, 'vol'))
+    source_morph_vol.save(os.path.join(tempdir, 'vol.h5'))
+
+    # check loading
+    source_morph_vol_r = read_source_morph(os.path.join(tempdir, 'vol.h5'))
+
+    # check for invalid file name handling
+    pytest.raises(IOError, read_source_morph, os.path.join(tempdir, '42'))
+
+    # check morph
+    stc_vol_morphed = source_morph_vol(stc_vol)
+
+    # check as_volume=True
+    assert isinstance(source_morph_vol(stc_vol, as_volume=True),
+                      nib.Nifti1Image)
```

Comment:
We should probably change to `Nifti2Image` as the default output in these new methods, see #5337

Reply:
it's an overkill to produce a volume to test output type.
Test type next to the line where you test this for correct output dim or whatever else you're testing.
THM -> try to be smart to avoid as much as possible testing time.